crates/hyperdrive-math/src/lp/remove.rs

```diff
@@ -1,6 +1,312 @@
+            };
+
+            // active_lp_total_supply and withdrawal_shares_total_supply are
+            // just the token supplies.  Get the from the contract.
+            let lp_token_asset_id = U256::zero();
+            let active_lp_total_supply: FixedPoint = bob
+                .hyperdrive()
+                .total_supply(lp_token_asset_id)
+                .await?
+                .into();
+            let withdrawal_share_asset_id = U256::from(3) << 248;
+            let withdrawal_shares_total_supply: FixedPoint = bob
+                .hyperdrive()
+                .total_supply(withdrawal_share_asset_id)
+                .await?
+                .into();
+
+            // Recod bob's shares and withdrawal shares before removing liquidity.
+            let starting_base = bob.wallet.base;
+            let starting_withdrawal_shares = bob.wallet.withdrawal_shares;
+
+            // Get the amount to remove, hit the budget 1% of the time to test
+            // that case but don't remove more than is possible.
+            let remove_budget = min(
+                rng.gen_range(fixed!(0)..=fixed!(1.01e18) * bob.wallet.lp_shares),
+                bob.wallet.lp_shares,
+            );
+            let remove_budget = min(
+                active_lp_total_supply - fixed!(2e18) * state.minimum_share_reserves(),
+                remove_budget,
+            );
+
+            // Bob removes liquidity
+            let tx_result = bob.remove_liquidity(remove_budget, None).await;
+
+            // Recod bob's base and withdrawal shares after removing liquidity.
+            let ending_base = bob.wallet.base;
+            let ending_withdrawal_shares = bob.wallet.withdrawal_shares;
+
+            // Calculate shares and withdrawal shares from the rust function.
+            let current_block_timestamp = bob.now().await?;
+            let result = std::panic::catch_unwind(|| {
+                state
+                    .calculate_remove_liquidity(
+                        current_block_timestamp,
+                        active_lp_total_supply,
+                        withdrawal_shares_total_supply,
+                        remove_budget,
+                        fixed!(0),
+                        fixed!(1),
+                    )
+                    .unwrap()
+            });
+
+            match result {
+                Ok((shares, withdrawal_shares)) => {
+                    // Assert base redeemed results match between rust and solidity
+                    // within a tolerance.
+                    let sol_base_result = I256::try_from(ending_base - starting_base)?;
+                    let vault_share_price = state.vault_share_price();
+                    let rust_base_reesult = I256::try_from(shares * vault_share_price)?;
+                    let base_diff =
+                        FixedPoint::try_from((sol_base_result - rust_base_reesult).abs())? / shares;
+                    assert!(base_diff < fixed!(0.0000001e18));
```

Comment:
I'm not sure why we need this. Shouldn't this be perfectly accurate?

Reply:
this is removed, exact now!